### PR TITLE
[Extensions] Update BaggageActivityProcessor to use baggage key predicate

### LINKED ISSUE #1695: Baggage activity processor - key predicate
I'd prefer if we had a constructor that allowed the user to add a prefix for baggage items that they'd like to add as attributes. We've had situations where customers have used a similar approach and managed to receive data from externally propagated baggage (accidentally).

A pattern like adding a prefix would mean that it becomes a lot more intentional, and therefore just a little safer.

_Originally posted by @martinjt in https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/1659#issuecomment-2076931103_

### PR DESCRIPTION
**Pull request by MikeGoldsmith**
Fixes #1695 

## Changes

- Updates the BaggageActivityProcessor to require a predicate as a constructor parameter
- Use the predicate function to determine which baggage entries are added to the new activity
- Add a public static predicate that allows all baggage entries
- Update README with examples of using the all baggage and a custom predicates

For significant contributions please make sure you have completed the following items:

* [x] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry.Extensions/Trace/BaggageActivityProcessor.cs
Comment (cijothomas) [inline src/OpenTelemetry.Extensions/Trace/BaggageActivityProcessor.cs:35]: I am not sure if running arbitrary user code run here is recommended.
Within OTel SDK, regex pattern was used instead:
Metrics : https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Metrics/MeterProviderSdk.cs#L154-L155 (allowing user to decide which meter names are enabled). Similar for Traces too.

Could we just accept a Regex pattern instead?

Comment (MikeGoldsmith) [reply src/OpenTelemetry.Extensions/Trace/BaggageActivityProcessor.cs:35]: I understand the concerns of running user-provided code. I'm trying super hard to be consistent across SDK implementations. Some want user-function, some want list of prefix strings and others want a regex. The user-function supports the other two.

Would you be okay keeping the user function if we add a regex extension method that takes a regex directly and add a warning to providing custom functions around being aware they are executed per activity?

Comment (cijothomas) [reply src/OpenTelemetry.Extensions/Trace/BaggageActivityProcessor.cs:35]: > custom functions around being aware they are executed per activity

It is per activity, for *each* baggage entry.

Comment (cijothomas) [reply src/OpenTelemetry.Extensions/Trace/BaggageActivityProcessor.cs:35]: I totally understand you are trying to get as much consistency as possible across languages, but the only sure way to make that happen is to have it come via Spec.

I'd prefer to *only* offer the Regex option here, it is consistent with how OTel .NET allows meter/activity-source. However, If other maintainers want to keep both, I won't block this PR!

Comment (cijothomas) [reply src/OpenTelemetry.Extensions/Trace/BaggageActivityProcessor.cs:35]: @open-telemetry/dotnet-contrib-maintainers Please share your thoughts.

Comment (MikeGoldsmith) [reply src/OpenTelemetry.Extensions/Trace/BaggageActivityProcessor.cs:35]: I'm not aware of spec that covers contrib components. If you can point me towards one I'd be happy to lobby it.

What are your concerns regarding providing the user-function? Security, performance, something else? In my mind, this is an opt-in processor that requires some level of configuration to select which baggage keys are selected. A regex is one way someone may want to configure it, but may not suit everyone.

I'm not dead-set on it being there, it's a goal to make the developer experience as consistent as possible. If it's a no-go from this SIG, I can deal with it.

Comment (MikeGoldsmith) [reply src/OpenTelemetry.Extensions/Trace/BaggageActivityProcessor.cs:35]: BTW we already use a user-function in the [AutoFlushActivityProcessor](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Extensions/Trace/AutoFlushActivityProcessor.cs#L41). It has instructions that function should not be complex (time-consuming) - would adding that warning be sufficient?

Comment (MikeGoldsmith) [reply src/OpenTelemetry.Extensions/Trace/BaggageActivityProcessor.cs:30]: That's my point, this is being added to contrib because we don't have a spec for it. My goal is to keep them consistent for developer experience as best as possible.

I'm not hard-set for it, but there is precedent for other contrib processors using a user-provided function.

**Comment by MikeGoldsmith** [comment]
Ping @open-telemetry/dotnet-contrib-approvers @open-telemetry/dotnet-contrib-maintainers - please can you take a look?

File: src/OpenTelemetry.Extensions/TracerProviderBuilderExtensions.cs
Comment (CodeBlanch) [inline src/OpenTelemetry.Extensions/TracerProviderBuilderExtensions.cs:83]: Is `*` a valid regex? This API seems a little awkward. Instead of forcing users to pass "*" as regex when they want everything why not just give them a default overload. Something like this...

```csharp
    // Captures all baggage
    public static TracerProviderBuilder AddBaggageActivityProcessor(this TracerProviderBuilder builder) {}

    // Captures baggage based on key predicate
    public static TracerProviderBuilder AddBaggageActivityProcessor(this TracerProviderBuilder builder, Predicate<string> baggageKeyPredicate) {}

    // Captures baggage based on key regex
    public static TracerProviderBuilder AddBaggageActivityProcessor(this TracerProviderBuilder builder, string baggageKeyRegex) {}
```

Personally, I wouldn't offer the regex overload. We have the `Predicate<string>` one users may implement that using a regex if they want, right?

File: src/OpenTelemetry.Extensions/TracerProviderBuilderExtensions.cs
Comment (CodeBlanch) [inline src/OpenTelemetry.Extensions/TracerProviderBuilderExtensions.cs:85]: Just FYI if you use the factory pattern it _should_ make these CA2000 warnings go away:

```csharp
return builder.AddProcessor(sp => new BaggageActivityProcessor(_ => true));
```

File: src/OpenTelemetry.Extensions/TracerProviderBuilderExtensions.cs
Comment (CodeBlanch) [inline src/OpenTelemetry.Extensions/TracerProviderBuilderExtensions.cs:67]: Just a random idea how to solve @cijothomas's concerns about user code throwing.

```csharp
return builder.AddProcessor(sp => new BaggageActivityProcessor(
   key => 
   {
        try
        {
            return baggageKeyPredicate(key);
        }
        catch (Exception ex)
        {
            // TODO: Log to EventSource
            return false;
        }
   }));
```

**Comment by MikeGoldsmith** [comment]
Thanks for the feedback @CodeBlanch - I've made some updates, please take another look.

File: src/OpenTelemetry.Extensions/Internal/OpenTelemetryExtensionsEventSource.cs
Comment (Kielek) [inline src/OpenTelemetry.Extensions/Internal/OpenTelemetryExtensionsEventSource.cs:51]:
```suggestion
    [Event(3, Message = "Baggage key predicate function threw exeption: '{0}'", Level = EventLevel.Error)]
    public void BaggageKeyPredicateException(string exception)
    {
        this.WriteEvent(3, exception);
```

File: src/OpenTelemetry.Extensions/README.md
Comment (cijothomas) [inline src/OpenTelemetry.Extensions/README.md:79]: could you also add that it is executed for each entry in the Baggage.

File: src/OpenTelemetry.Extensions/Trace/BaggageActivityProcessor.cs
Comment (cijothomas) [inline src/OpenTelemetry.Extensions/Trace/BaggageActivityProcessor.cs:12]: is this needed to be public?

File: src/OpenTelemetry.Extensions/Internal/OpenTelemetryExtensionsEventSource.cs
Comment (cijothomas) [inline src/OpenTelemetry.Extensions/Internal/OpenTelemetryExtensionsEventSource.cs:48]:
```suggestion
    [Event(3, Message = "Baggage key predicate function threw exeption: '{0}'", Level = EventLevel.Warning)]
```

Comment (cijothomas) [reply src/OpenTelemetry.Extensions/Internal/OpenTelemetryExtensionsEventSource.cs:48]: Error is better suited for more serious conditions like telemetry loss etc. This one is less severe, as the telemetry will continue to flow without issues, just that the baggage won't be attached.

Comment (cijothomas) [reply src/OpenTelemetry.Extensions/Internal/OpenTelemetryExtensionsEventSource.cs:48]: Also - could we make the message more clear, to indicate the impact of this? Will this particular key be not added? or entire baggage?

Comment (MikeGoldsmith) [reply src/OpenTelemetry.Extensions/Internal/OpenTelemetryExtensionsEventSource.cs]: Updated message in [bc08c17](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1816/commits/bc08c17ae39a79a6b87c282d6e109ad151e47142)

File: src/OpenTelemetry.Extensions/Trace/BaggageActivityProcessor.cs
Comment (MikeGoldsmith) [reply src/OpenTelemetry.Extensions/Trace/BaggageActivityProcessor.cs:12]: It needs to be public to expose the `AllowAllBaggageKeys` convenience predicate. I did originally have it in  `TracerProviderBuilderExtensions` but felt weird referencing the predicate as `TracerProviderBuilderExtensions.AllowAllBaggageKeys` and using that class directly resulted in namespace conflicts as the class name exists in a number of other packages.

Adding to `BaggageActivityProcessor` feels nicer as you then reference the predicate as `BaggageActivityProcessor.AllowAllBaggageKeys`.

I'm open alternatives, I don't like I've had to add the class and other parts to the public API.

File: src/OpenTelemetry.Extensions/README.md
Comment (MikeGoldsmith) [reply src/OpenTelemetry.Extensions/README.md]: Update in [f147ef9](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1816/commits/f147ef9b1b9c803c48254a4302adc007e8ef92b4)

File: src/OpenTelemetry.Extensions/Trace/BaggageActivityProcessor.cs
Comment (cijothomas) [reply src/OpenTelemetry.Extensions/Trace/BaggageActivityProcessor.cs:12]: got it. as long as we prevent users from doing `new BaggageActivityProcessor()` (which would be incorrect), this is fine.

File: test/OpenTelemetry.Extensions.Tests/Trace/BaggageActivityProcessorTests.cs
Comment (cijothomas) [inline test/OpenTelemetry.Extensions.Tests/Trace/BaggageActivityProcessorTests.cs:19]: not sure what is the need of `TestActivityProcessor` in this test? Most likely a leftover from other tests?

File: src/OpenTelemetry.Extensions/Internal/OpenTelemetryExtensionsEventSource.cs
Comment (cijothomas) [inline src/OpenTelemetry.Extensions/Internal/OpenTelemetryExtensionsEventSource.cs:48]: something is wrong with this error message..
It is not clear if all baggage entries will be dropped or a particular baggageentry will be dropped ?

File: test/OpenTelemetry.Extensions.Tests/Trace/BaggageActivityProcessorTests.cs
Comment (cijothomas) [inline test/OpenTelemetry.Extensions.Tests/Trace/BaggageActivityProcessorTests.cs:97]: I'd suggest a test to validate the behavior - when an exception is thrown, is entire baggage dropped? or just that particular baggage entry is dropped? If we are dropping only the particular baggage entry, then the error message should contain the keyname.

File: test/OpenTelemetry.Extensions.Tests/Trace/BaggageActivityProcessorTests.cs
Comment (MikeGoldsmith) [reply test/OpenTelemetry.Extensions.Tests/Trace/BaggageActivityProcessorTests.cs]: Removed in [19fda8c](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1816/commits/19fda8cc0eea9d6d9af678672b8199455d9a1c9c)

File: src/OpenTelemetry.Extensions/Internal/OpenTelemetryExtensionsEventSource.cs
Comment (MikeGoldsmith) [reply src/OpenTelemetry.Extensions/Internal/OpenTelemetryExtensionsEventSource.cs]: Updated message with baggage key in [87478e0](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1816/commits/87478e0c629af630e1aaf7a18ad0e497ebcf3f7a)

File: test/OpenTelemetry.Extensions.Tests/Trace/BaggageActivityProcessorTests.cs
Comment (MikeGoldsmith) [reply test/OpenTelemetry.Extensions.Tests/Trace/BaggageActivityProcessorTests.cs:88]: Added extra test to show only entries that throw are dropped in [325f39f](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1816/commits/325f39f0836150f163c088451fe13de2b69ba37f)

File: test/OpenTelemetry.Extensions.Tests/Trace/BaggageActivityProcessorTests.cs
Comment (MikeGoldsmith) [reply test/OpenTelemetry.Extensions.Tests/Trace/BaggageActivityProcessorTests.cs]: Fixed in [3239b2e](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1816/commits/3239b2ed4fc66509c29d38caa67abf5be3666cb0)

**Review by cijothomas — APPROVED** [review]
I am generally not in favor of allowing arbitrary user code executions (see https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1816#discussion_r1607213069), but given there is an existing precedent in **this** component, not a blocker.

**Comment by Kielek** [comment]
@MikeGoldsmith, could you please merge main branch to PR branch (or allow us to do this)?

@CodeBlanch, do you have any other comments?

**Comment by MikeGoldsmith** [comment]
@Kielek updated again, let me know if you need anything else.

**Comment by cijothomas** [comment]
> @Kielek updated again, let me know if you need anything else.

Can you fix the CI failures? Looks like some minor import statements causing CI to fail.

**Comment by MikeGoldsmith** [comment]
Yeah, sure. I wasn't sure if they were intermittent / unrelated failures as they only occurred in the merge from main commits. I'll fix up the PR tomorrow.

**Comment by Kielek** [comment]
@MikeGoldsmith, it is the reason #1927 (or one of the previous PR for the same topic), unfortunately needs to be manually adjusted for all existing PRs.

**Pull request merged**
